### PR TITLE
Remove unnecessary allocations and clones

### PR DESCRIPTION
**Pull request by vbrandl**
Another PR by me :)

When working on my last change, I saw a few unnecessary `clone`s, `to_string`s, `to_owned`s and so on. This change tries to remove them where possible and therefore improve the runtime performance by getting rid of a few heap allocations and memory copies.

There are no new features, so I didn't write new tests, but the existing tests still pass :)

#### ✔ Checklist:

- [x] Commit messages have been written in [Conventional Commits](https://www.conventionalcommits.org) format;
- [x] This PR has been added to [CHANGELOG.md](https://github.com/dotenv-linter/dotenv-linter/blob/master/CHANGELOG.md) (at the top of the list);
- [ ] Tests for the changes have been added (for bug fixes / features);
- [ ] Docs have been added / updated on the [dotenv-linter.github.io](https://github.com/dotenv-linter/dotenv-linter.github.io) (for bug fixes / features).

### CODE REVIEW

**Comment by vbrandl** [comment]
I'm not familiar with codecov. Is there a way to see, which branch was covered by testing before, and is not covered after this change?

**Comment by vbrandl** [comment]
1a5e70a is a personal preference. I like the iterator API better than `for` loops. If you are unhappy with any of the changes, I'll revert those.

**Comment by mgrachev** [comment]
> [1a5e70a](https://github.com/dotenv-linter/dotenv-linter/commit/1a5e70a4e188602cccd3647a1c0daf5c57ff634b) is a personal preference. I like the iterator API better than `for` loops. If you are unhappy with any of the changes, I'll revert those.

I like this approach too 🙂

File: src/common/warning.rs [hunk 1]
```diff
@@ -9,9 +9,10 @@ pub struct Warning {
     message: String,
 }
 
-impl Warning {
-    pub fn new(line: LineEntry, check_name: &str, message: String) -> Self {
-        let check_name = String::from(check_name);
+impl<'a> Warning {
```
[the remark this answers is not in the document]
Comment (vbrandl) [reply src/common/warning.rs]: It does most definitely not. Just a artifact from when I tried to refactor `Warning` so it does not actually own any values but just takes borrows to `LineEntry`, `check_name` and `message`, but this didn't work as intended. I removed it again.

**Comment by mgrachev** [comment]
> I'm not familiar with codecov. Is there a way to see, which branch was covered by testing before, and is not covered after this change?

You can see the diff here https://app.codecov.io/gh/dotenv-linter/dotenv-linter/compare/350/diff

**Comment by mgrachev** [comment]
@vbrandl Thank you for your contribution! 🔥

> - [ ] This PR has been added to [CHANGELOG.md](https://github.com/dotenv-linter/dotenv-linter/blob/master/CHANGELOG.md) (at the top of the list);

Please do it 🙂

**Comment by vbrandl** [comment]
> > I'm not familiar with codecov. Is there a way to see, which branch was covered by testing before, and is not covered after this change?
> 
> You can find the diff here https://app.codecov.io/gh/dotenv-linter/dotenv-linter/compare/350/diff

I still cannot really tell, which branches, that were covered before, are not covered now, so I know for which cases, new tests are required.

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 ### 🚀 Added
+- Refactoring to get rid of unnecessary heap allocations and some improvements in API ergonomics [#350](https://github.com/dotenv-linter/dotenv-linter/pull/350) ([@vbrandl](https://github.com/vbrandl))
```
Comment (mgrachev) [inline CHANGELOG.md:9]: Please move this line to `Changed` section.

Comment (vbrandl) [reply CHANGELOG.md]: Sorry missed that

**Review by mgrachev — APPROVED** [review]

**Comment by vbrandl** [comment]
It might actually be interesting to rerun the benchmarks from #351 to see, if these changes have a measurable impact

**Comment by mgrachev** [comment]
@vbrandl Thanks a lot 🚀 You did a great job 👍

**Comment by mgrachev** [comment]
> It might actually be interesting to rerun the benchmarks from #351 to see, if these changes have a measurable impact

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `target/release/dotenv-linter .env` | 2.7 ± 0.4 | 2.0 | 4.3 | 1.00 |
| `/Users/mgrachev/Library/Python/3.8/bin/dotenv-linter .env` | 162.6 ± 12.1 | 153.0 | 201.3 | 60.83 ± 10.20 |

**Comment by vbrandl** [comment]
I count cutting down the max execution time by almost 50% as a win, even if it's indistinguishable for humans :D

**Comment by mgrachev** [comment]
> I count cutting down the max execution time by almost 50% as a win, even if it's indistinguishable for humans :D

You are awesome ⚡️

**Pull request merged**